### PR TITLE
[DOC] Adding architecture documentation to explanations

### LINKED ISSUE #11: [DOC] Nebari architecture and conceptual guide
Add documentation page (md) titled `Nebari architecture and conceptual guide`

👉  [notion page](https://www.notion.so/quansightlabs/Nebari-architecture-and-conceptual-guide-05a50bd693ae4e98b166833114896faa)

Examples in other documentation sites
- https://kubernetes.io/docs/concepts/
- https://www.kubeflow.org/docs/started/architecture/

### LINKED ISSUE #27: [DOC] -  Nebari multi-cloud architectural diagrams
We are lacking in-depth / up-to-date architectural diagrams

- [ ] #28
- [ ] #29
- [ ] #30
- [ ] #31

Diagrams should be added to https://lucid.app/documents#/documents?folder_id=278399173

### LINKED ISSUE #28: [DOC] - Architectural diagram Azure

### LINKED ISSUE #29: [DOC] - Architectural diagram GCP

### LINKED ISSUE #30: [DOC] - Architectural diagram DO

### LINKED ISSUE #31: [DOC] - Architectural diagram AWS

### PR DESCRIPTION
**Pull request by costrouc**
Closes #11 
Closes #27 
Closes #28 
Closes #29 
Closes #30 
Closes #31

This is still a work in progress.

### CODE REVIEW

File: docs/docs/explanations/architecture.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+### Azure
+
+![Azure Architecture Diagram](/img/architecture-diagram-azure.png)
+
+## Kubernetes services
+
+![Kubernetes Architecture Diagram](/img/architecture-diagram-kubernetes.png)
+
+The most significant effort in Nebari is deploying the services within
+the Kubernetes cluster. All of the infrastructure diagrams before
```
Comment (kcpevey) [inline docs/docs/explanations/architecture.md:47]: i'm beginning to realize maybe you meant the diagrams were ready for review, not the PR itself...

File: docs/docs/explanations/architecture.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+limitation of Digital Ocean auto-scaling is that it only scales down
+to one. Once Nebari is fully deployed a load-balancer is connected to
+the Kubernetes cluster through [Traefik](https://traefik.io/).
+
+### Amazon Web Services (AWS)
+
+![AWS Architecture Diagram](/img/architecture-diagram-aws.png)
+
+AWS is the most complex of the cloud deployments with significant
+effort being put into networking. An EKS managed kubernetes cluster is
```
[the remark this answers is not in the document]
Comment (kcpevey) [reply docs/docs/explanations/architecture.md:29]: also inconsistency on capitalization of kubernetes.

File: docs/docs/references/architecture.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+called by the user. It is automatically run with the `deploy`
+operation. A `stages` directory is produced.
+
+## [nebari/deploy.py](https://github.com/Quansight/qhub/blob/main/qhub/deploy.py) 
+
+`deploy` infrastructure and kubernetes resources via [Terraform](https://www.terraform.io/)
+
+## [nebari/destroy.py](https://github.com/Quansight/qhub/blob/main/qhub/destroy.py) 
+
+`destroy` infrastructure and kubernetes resources via [Terraform](https://www.terraform.io/). The `destroy` command roughly does the `deploy` command in reverse.
```
Comment (kcpevey) [inline docs/docs/references/architecture.md:32]:
```suggestion
The `destroy` command roughly does the `deploy` command in reverse.
```

**Review by kcpevey — DISMISSED** [review]

**Comment by trallard** [comment]
ping @costrouc  - there are changes requested. Also, where are we up to with the new architecture diagrams?

File: docs/docs/explanations/architecture.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+cloud specific infrastructure diagrams along with a single diagram to
+outline the Kubernetes services.
+
+## Infrastructure
+
+### Digital Ocean (DO)
+
+The Digital Ocean deployment is the simplest of the cloud
+providers. We deploy a managed Kubernetes not within a virtual
+network. Similar to all other deployments auto-scaling is enabled. A
```
[the remark this answers is not in the document]
Comment (viniciusdc) [reply docs/docs/explanations/architecture.md:19]: I will redo all text contents, but answering your question is more to the first option: We deploy a managed Kubernetes, and its not within a virtual network

**Comment by viniciusdc** [comment]
@kcpevey I will mark your suggestions above as "resolved" as I plan to re-write everything, I will keep them in mind though.

File: docs/docs/explanations/architecture.mdx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+Some of the most important concepts that Nebari relies on are Terraform and Kubernetes.
+
+![Nebari core concepts](/img/explanations/core-concepts.png)
+
+## Terraform
+
+Terraform is an excellent option for deploying and managing cloud
+infrastructure. It allows you to easily configure any number of
+cloud servers as well as all the required connections and features
+to ensure security, redundancy and scalability
```
Comment (aktech) [inline docs/docs/explanations/architecture.mdx:20]: full stop after "scalability"

File: docs/docs/explanations/architecture.mdx [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+etc. Such methodology is usualy referred to as Infrastructure as code (IaC).
+
+For more information about Terraform, IaC and the features it provides, refer to
+[Infrastructure as Code with Terraform](https://learn.hashicorp.com/tutorials/terraform/infrastructure-as-code).
+
+![Terraform](/img/explanations/terraform.png)
+
+## Kubernetes
+
+Kubernetes is a powerful open-source system, initially developed by Google,
```
Comment (aktech) [inline docs/docs/explanations/architecture.mdx:35]: "open-source system" -> "open-source container orchestration system"

File: docs/docs/explanations/architecture.mdx [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+for managing containerized applications in a clustered environment. It aims
```
Comment (aktech) [inline docs/docs/explanations/architecture.mdx:36]: Not sure if "clustered environment" is a commonly used term. I would suggest `cluster` instead.

File: docs/docs/explanations/architecture.mdx [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+to provide better ways of managing related, distributed components and services
+across varied infrastructure.
+
+Kubernetes provides fine-grained control in applications and container lifecycles,
```
Comment (aktech) [inline docs/docs/explanations/architecture.mdx:40]: control -> controls

File: docs/docs/explanations/architecture.mdx [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+especially in dynamic environments. Tasks such as pod tracking, availability, and
```
Comment (aktech) [inline docs/docs/explanations/architecture.mdx:41]: I wouldn't write "pod tracking" here. We're explaining Kubernetes here and assuming people would know what's a pod is probably not a good idea and "pod tracking" isn't really a popular term anyway.

**Comment by trallard** [comment]
@costrouc what is needed to move this PR forward? it has been in progress for a while and it would be great to add it to the docs ASAP

**Comment by kcpevey** [comment]
It was originally owned by @costrouc, then @viniciusdc took over. It definitely needs to be merged ASAP.

**Comment by viniciusdc** [comment]
Based on what has been discussed so far and the reviews. We could avoid describing the diagrams in detail (no details at all for now) and refactor the page layout to show the images. We can then have another issue to update them at a later time. This way, we can move forward with this.

Currently, the missing thing on this is that we only added the images but did not explain where these resources are used or how they are used (a.ka Jupyterlab pods are initiated by a user trigger that spawns a Pod with the requested resources.... and is scaled down once idle)

**Comment by viniciusdc** [comment]
> It was originally owned by @costrouc, then @viniciusdc took over. It definitely needs to be merged ASAP.

Based on what has been discussed so far and the reviews. We could avoid describing the diagrams in detail (no details at all for now) and refactor the page layout to show the images. We can then have another issue to update them at a later time. This way, we can move forward with this.

Currently, the missing thing on this is that we only added the images but did not explain where these resources are used or how they are used (a.ka Jupyterlab pods are initiated by a user trigger that spawns a Pod with the requested resources.... and is scaled down once idle)

File: docs/docs/explanations/architecture.mdx [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+Some of the most important concepts that Nebari relies on are Terraform and Kubernetes.
+
+![Nebari core concepts](/img/explanations/core-concepts.png)
+
+## Terraform
+
+Terraform is an excellent option for deploying and managing cloud
+infrastructure. It allows you to easily configure any number of
+cloud servers as well as all the required connections and features
+to ensure security, redundancy and scalability
```
Comment (pavithraes) [inline docs/docs/explanations/architecture.mdx:20]:
```suggestion
to ensure security, redundancy, and scalability.
```

File: docs/docs/explanations/architecture.mdx [hunk 11]
```diff
@@ -0,0 +1,70 @@
+---
+id: infrastructure-architecture
+title: Nebari architecture and conceptual guide
```
Comment (pavithraes) [inline docs/docs/explanations/architecture.mdx:3]:
```suggestion
title: Nebari architecture
```

Comment (pavithraes) [reply docs/docs/explanations/architecture.mdx:3]: Context: The documentation section where this page lives is called "Conceptual guides", so we needn't say it again explicitely :)

File: docs/docs/explanations/architecture.mdx [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+Kubernetes is a powerful open-source container orchestration system, initially developed by Google,
+for managing containerized applications in a cluster. It aims
+to provide better ways of managing related, distributed components and services
+across varied infrastructure.
+
+Kubernetes provides fine-grained controls in applications and container lifecycles,
+especially in dynamic environments. Tasks such as monitoring, availability, and
+deployment are simplified by the use of controllers, which are control loops that
+monitor your Kubernetes cluster and make changes as needed to maintain the expected
+state.
```
Comment (pavithraes) [inline docs/docs/explanations/architecture.mdx:35]:
```suggestion
state.
You can learn more details in the [Kubernetes documentation](https://kubernetes.io/docs/concepts/overview/).
```

File: docs/docs/explanations/architecture.mdx [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+![Kubernetes overall structure levels](/img/explanations/kubernetes-concepts-architecture.jpg)
+
+## Infrastructure
+
+Below are diagrams to help show the architecture and design behind
+Nebari. The diagrams are meant to show different levels. There are
+cloud specific infrastructure diagrams along with a single diagram to
+outline the Kubernetes services.
```
Comment (pavithraes) [inline docs/docs/explanations/architecture.mdx:44]: I believe the "Kubernetes services" diagram is the one in the previous section?

```suggestion
Below are diagrams to help show the architecture and design behind
Nebari. The diagrams are meant to show different levels and cloud specific infrastructure.
```

**Review by pavithraes — APPROVED** [review]
(deleted, clicked by approve by mistake!)

**Review by pavithraes — CHANGES_REQUESTED** [review]
@viniciusdc Thank you for taking over this PR and updating it! It's looking good!

I've shared some small suggestions inline, and have a couple of high-level comments:

- We need to merge the main branch again, and update sidebar.js -- it currently shows "References" twice.
- We can remove the images that aren't being used from this PR.

**Comment by viniciusdc** [comment]
Self-review: I will remove the unused images from this PR as well as mention the source of one of the terraform images as it comes from their docs (thus needs citation)

**Review by pavithraes — APPROVED** [review]
Thanks for the updates, `@viniciusdc`!

**Pull request merged**